File: server/routers/admin/src/product_management/products/schema.rs
use entity::{
    product::{ ProductHighlights, ProductImages, ProductSeo, ProductVideo },
    sea_orm_active_enums::ProductStatus,
};
use sea_orm::prelude::Decimal;
use serde::{ Deserialize, Serialize };
use utoipa::ToSchema;
use validator::Validate;

#[derive(Debug, ToSchema, Validate, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateProductInput {
    #[validate(
        length(min = 1, message = "Title is required"),
        length(max = 200, message = "Maximum 200 characters are allowed")
    )]
    #[schema(example = "Color", min_length = 1, max_length = 50)]
    /// Title of the product.
    ///
    pub title: String,
    pub description: serde_json::Value,
    #[validate(length(min = 1, message = "Category is required"))]
    pub category: String,
    #[validate(length(min = 1, message = "Brand is required"))]
    pub brand: String,
    #[validate(length(min = 1, message = "Color is required"))]
    pub color: String,
    #[validate(length(min = 1, message = "Size is required"))]
    pub size: String,
    #[validate(length(min = 1, message = "Style is required"))]
    pub style: String,
    pub highlights: ProductHighlights,
    pub images: ProductImages,
    pub is_returnable: bool,
    pub maximum_order: i32,
    pub minimum_order: i32,
    pub mrp: Decimal,
    pub price: Decimal,
    pub seo: ProductSeo,
    #[validate(length(min = 1, message = "SKU ID is required"))]
    pub sku_id: String,
    pub stock: i32,
    pub tags: Vec<String>,
    pub video: ProductVideo,
    pub status: ProductStatus,
}

#[derive(Debug, ToSchema, Serialize)]
pub struct ProductWithMessage {
    pub message: String,
    pub data: entity::product::Model,
}

File: client/apps/admin/src/app/(dashboard)/product-management/products/create/_lib/schema.ts
import { ProductStatus, type CreateProductInput } from "@/client";
import {
  coerceNumber,
  nativeEnum,
  string,
  withEmptyString,
  withSchema,
  zodImage,
} from "@repo/utils/zod";
import { any, array, boolean, number, record, set } from "zod";

const highlights = withSchema<CreateProductInput["highlights"][number]>()({
  description: withEmptyString(string().min(1, "Description is required")),
  highlight: string().min(1, "Highlight is required"),
});

const video = withSchema<CreateProductInput["video"]>()({
  url: withEmptyString(string().min(1, "Url is required")),
  thumbnail: zodImage("Thumbnail is required"),
});

const seo = withSchema<CreateProductInput["seo"]>()({
  title: withEmptyString(string().min(1, "Title is required")),
  description: withEmptyString(string().min(1, "Description is required")),
  keywords: array(string().min(1, "Keyword is required")),
});

export const productCreateSchema = withSchema<CreateProductInput>()({
  brand: string().min(1, "Brand is required"),
  category: string().min(1, "Category is required"),
  color: string().min(1, "Color is required"),
  description: record(any()),
  highlights: array(highlights),
  images: array(zodImage("Images are required")).min(
    4,
    "Minimum 4 images are required",
  ),
  isReturnable: boolean(),
  maximumOrder: coerceNumber(number().positive(), "Maximum order is required"),
  minimumOrder: coerceNumber(number().positive(), "Maximum order is required"),
  mrp: coerceNumber(number().positive(), "MRP is required"),
  price: coerceNumber(number().positive(), "Price is required"),
  stock: coerceNumber(number().positive(), "Stock is required"),
  size: string().min(1, "Size is required"),
  style: string().min(1, "Style is required"),
  title: string().min(1, "Title is required"),
  status: nativeEnum(ProductStatus, "Status is required"),
  tags: set(string().min(1, "Tag is required")).transform((val) => [...val]),
  skuId: string().min(1, "Sku id is required"),
  video,
  seo,
});

productCreateSchema._type satisfies CreateProductInput;
